src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1857 @@
+        if (notify_fd < 0)
+                return -errno;
+
+        (void) mkdir_parents_label(sa.un.sun_path, 0755);
+        (void) sockaddr_un_unlink(&sa.un);
+
+        if (bind(notify_fd, &sa.sa, SOCKADDR_UN_LEN(sa.un)) < 0)
+                return -errno;
+
+        r = setsockopt_int(notify_fd, SOL_SOCKET, SO_PASSCRED, true);
+        if (r < 0)
+                return r;
+
+        r = sd_event_add_io(m->event, &m->notify_event, notify_fd, EPOLLIN, manager_on_notify, m);
+        if (r < 0)
+                return r;
+
+        (void) sd_event_source_set_description(m->notify_event, "notify-socket");
+
+        r = sd_event_source_set_io_fd_own(m->notify_event, true);
+        if (r < 0)
+                return r;
+        TAKE_FD(notify_fd);
+
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+static int manager_enumerate_image_class(Manager *m, TargetClass class) {
+        _cleanup_hashmap_free_ Hashmap *images = NULL;
+        Image *image;
+        int r;
+
+        images = hashmap_new(&image_hash_ops);
+        if (!images)
+                return log_oom();
+
+        r = image_discover((ImageClass) class, NULL, images);
+        if (r < 0)
+                return r;
+
+        HASHMAP_FOREACH(image, images) {
+                _cleanup_(target_freep) Target *t = NULL;
+                bool have = false;
+
+                if (IMAGE_IS_HOST(image))
+                        continue; /* We already enroll the host ourselves */
+
+                r = target_new(m, class, image->name, image->path, &t);
+                if (r < 0)
+                        return r;
+                t->image_type = image->type;
+
+                r = target_list_components(t, NULL, &have);
+                if (r < 0)
+                        return r;
+                if (!have) {
+                        log_debug("Skipping %s because it has no default component", image->path);
+                        continue;
+                }
+                TAKE_PTR(t); /* Now owned by m->targets */
+        }
+
+        return 0;
+}
+
+static int manager_enumerate_components(Manager *m) {
+        _cleanup_strv_free_ char **components = NULL;
+        bool have_default;
+        Target *t;
+        int r;
+
+        r = target_list_components(NULL, &components, &have_default);
+        if (r < 0)
+                return r;
+
+        if (have_default) {
+                r = target_new(m, TARGET_HOST, "host", "sysupdate.d", &t);
+                if (r < 0)
+                        return r;
+        }
+
+        STRV_FOREACH(component, components) {
+                _cleanup_free_ char *path = NULL;
+
+                path = strjoin("sysupdate.", *component, ".d");
+                if (!path)
+                        return log_oom();
+
+                r = target_new(m, TARGET_COMPONENT, *component, path, &t);
+                if (r < 0)
+                        return r;
+        }
+
+        return 0;
```

Comment:
so, in systemd we want that functions either log about all the errors they encounter, or don't (in which case it's up to the caller to log if it wants). But it should be either/or, i.e. log about all errors or about none. Hence for each function you have to decide which kind it should be. usually "main-function-like" stuff should log, and "library-like" stuff should not. But there's a blurry line between the two. 

(debug logging is always ok, btw, it doesn't count as "logging" in this sense, only LOG_INFO and up is considered logging)

the goal thus ultimately is that we log about all errors once, but not more than once.

This function here, will log in the strjoin() failure, but not on the target_new() case, which is why i am saying all this. please clean this up here (and everywhere else)

Functions that implement method handlers for IPC should typically log on their own if they encountered errors that are not the client's immediate faults. If we run into errors with client's passed params just return an error to the client and be done.

Reply:
Ditto with `target_list_components`.

Question for Lennart: Let's say, for example, that `target_new` did its own logging. Then we could just `return r` and be done, right, to avoid double-logging?

Reply:
correct.